server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -1372,12 +1373,18 @@ private boolean indexValue(DocumentParserContext context, XContentString value)
 
         if (fieldType().storedInBinaryDocValues()) {
             assert fieldType.docValuesType() == DocValuesType.NONE;
-            MultiValuedBinaryDocValuesField field = (MultiValuedBinaryDocValuesField) context.doc().getField(fieldType().name());
+
+            var field = (MultiValuedBinaryDocValuesField.SeparateCount) context.doc().getField(fieldType().name());
```

Comment:
```suggestion
            var field = (MultiValuedBinaryDocValuesField.SeparateCount) context.doc().getByKey(fieldType().name());
```

`getByField` does a linear scan of all fields in the document and is much slower than `getByKey` which uses a map.